tlaplus/dynamodblock.tla

```diff
@@ -0,0 +1,633 @@
+\* Parameter record_version of procedure DynamodbUpsertLock at line 73 col 5 changed to record_version_
+VARIABLES shared_counter, time_now, active_writer, local_record_version, 
+          local_counter, re_upsert_err, lock_owner, lock_released, 
+          lock_record_version, lock_duration, pc, stack
+
+(* define statement *)
+SumSet(set) ==
+    LET F[s \in SUBSET set] ==
+        IF s = {}
+        THEN 0
+        ELSE
+            LET x == CHOOSE x \in s : TRUE
+            IN x + F[s \ {x}]
+    IN F[set]
+
+
+StructValues(st) == {st[x] : x \in DOMAIN st}
+
+
+ConsistentCount == shared_counter >= SumSet(StructValues(local_counter))
+
+NoMissCount == <>[](shared_counter = Cardinality(WRITERS) * LOOP_COUNT)
+
+VARIABLES owner_, record_version_, cached_lock_lookup_time, 
+          cached_lock_duration, owner, record_version, next_time_tick, 
+          cached_lock_lookup_time_, cached_lock_duration_, 
+          cached_lock_version, local_count_value
+
+vars == << shared_counter, time_now, active_writer, local_record_version, 
+           local_counter, re_upsert_err, lock_owner, lock_released, 
+           lock_record_version, lock_duration, pc, stack, owner_, 
+           record_version_, cached_lock_lookup_time, cached_lock_duration, 
+           owner, record_version, next_time_tick, cached_lock_lookup_time_, 
+           cached_lock_duration_, cached_lock_version, local_count_value >>
+
+ProcSet == (WRITERS) \cup {TIME_TICK_UNIT}
+
+Init == (* Global variables *)
+        /\ shared_counter = 0
+        /\ time_now = 0
+        /\ active_writer = Cardinality(WRITERS)
+        /\ local_record_version = [ x \in WRITERS |-> 0 ]
+        /\ local_counter = [ x \in WRITERS |-> 0 ]
+        /\ re_upsert_err = [ x \in WRITERS |-> NULL ]
+        /\ lock_owner = NULL
+        /\ lock_released = NULL
+        /\ lock_record_version = NULL
+        /\ lock_duration = NULL
+        (* Procedure DynamodbUpsertLock *)
+        /\ owner_ = [ self \in ProcSet |-> NULL]
+        /\ record_version_ = [ self \in ProcSet |-> NULL]
+        /\ cached_lock_lookup_time = [ self \in ProcSet |-> NULL]
+        /\ cached_lock_duration = [ self \in ProcSet |-> NULL]
+        (* Procedure AcquireLock *)
+        /\ owner = [ self \in ProcSet |-> NULL]
+        /\ record_version = [ self \in ProcSet |-> NULL]
+        /\ next_time_tick = [ self \in ProcSet |-> NULL]
+        /\ cached_lock_lookup_time_ = [ self \in ProcSet |-> NULL]
+        /\ cached_lock_duration_ = [ self \in ProcSet |-> NULL]
+        /\ cached_lock_version = [ self \in ProcSet |-> NULL]
+        (* Procedure CriticalSection *)
+        /\ local_count_value = [ self \in ProcSet |-> NULL]
+        /\ stack = [self \in ProcSet |-> << >>]
+        /\ pc = [self \in ProcSet |-> CASE self \in WRITERS -> "writerloop"
+                                        [] self = TIME_TICK_UNIT -> "tick"]
+
+cond_put(self) == /\ pc[self] = "cond_put"
```

Comment:
![ohgod](http://3.bp.blogspot.com/-uuR-fhXKHuI/UQi_Hi7J3_I/AAAAAAAABq0/Ms4HcB-iffk/s320/my+eyes.jpg)


This looks like Pascal and LaTeX joined forces with brainfuck

Reply:
Yeah, it's heavily inspired by LaTex, one of the features it provides is rendering the spec in a pdf just like latex :(